File: app/src/main/java/com/example/farith/dailynotes/NoteActivity.java
package com.example.farith.dailynotes;

import android.app.AlarmManager;
import android.app.AlertDialog;
import android.app.DatePickerDialog;
import android.app.PendingIntent;
import android.app.TimePickerDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.database.sqlite.SQLiteDatabase;
import android.os.SystemClock;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.TimePicker;
import android.widget.Toast;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Random;

public class NoteActivity extends AppCompatActivity {
    EditText noteText;
    String notes;
    String adapterPosition;
    String previousTime;
    Button saveBtn;
    NoteDb finalDatabase;
    SQLiteDatabase database;
    private static final String TAG = NoteActivity.class.getSimpleName();
    Button btnRemindMe;
    String formattedDate;
    String notificationID;
    String reminderTime = "No reminder Added";
    TextView reminderText;
    BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String cancelNotificationId = intent.getStringExtra("cancel_noti");
            Log.d(TAG, "onReceive: cancelling notification id is");
            Intent newIntent = new Intent(NoteActivity.this, Notification.class);
            PendingIntent pendingIntent = PendingIntent.getBroadcast(getApplicationContext(), Integer.parseInt(cancelNotificationId), newIntent, PendingIntent.FLAG_UPDATE_CURRENT);
            AlarmManager alarmManager = (AlarmManager) getSystemService(ALARM_SERVICE);
            alarmManager.cancel(pendingIntent);
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_note);
        reminderText = findViewById(R.id.reminder_text);
        noteText = findViewById(R.id.note_text);
        saveBtn = findViewById(R.id.save_btn);
        btnRemindMe = findViewById(R.id.btn_remind_me);
        finalDatabase = new NoteDb(this);
        saveBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //making sure the edit text is not empty
                if (TextUtils.isEmpty(noteText.getText().toString().trim())) {
                    Toast.makeText(NoteActivity.this, "the Note is empty", Toast.LENGTH_SHORT).show();
                    Intent backToMainActivity = new Intent(NoteActivity.this, MainActivity.class);
                    backToMainActivity.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(backToMainActivity);
                } else {
                    //Checking whether the intent is from adapter or add button by verifying the intent extra
                    if (!TextUtils.isEmpty(adapterPosition)) {
                        Log.d(TAG, "onClick: save button is clicked " + noteText.getText().toString() + " " + adapterPosition);
                        finalDatabase.updateDb(database, noteText.getText().toString(), previousTime, getTimeInMillisconds(), notificationID, reminderTime);
                        Intent backToMainActivity = new Intent(NoteActivity.this, MainActivity.class);
                        backToMainActivity.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                        startActivity(backToMainActivity);
                    } else {
                        notificationID = createNotificationId();
                        Log.d(TAG, "onClick: notification id created and the id is " + notificationID);
                        finalDatabase.insertData(database, noteText.getText().toString(), getTimeInMillisconds(), notificationID, reminderTime);
                        Intent backToMainActivity = new Intent(NoteActivity.this, MainActivity.class);
                        backToMainActivity.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                        startActivity(backToMainActivity);
                    }
                }
            }
        });
        Intent intent = getIntent();
        notes = intent.getStringExtra("notes");
        adapterPosition = intent.getStringExtra("position");
        previousTime = intent.getStringExtra("previousTime");
        notificationID = intent.getStringExtra("noti");
        reminderTime = intent.getStringExtra("reminder_time");
        Log.d(TAG, "onCreate:notes " + notes + "onCreate: adapterPosition " + adapterPosition + " reminder time from intent " + reminderTime);
        if (TextUtils.isEmpty(notes)) {

        } else {
            noteText.setText(notes);
            //This line is to take the cursor to the end of the edit text field
            noteText.setSelection(noteText.getText().toString().length());
            if (TextUtils.isEmpty(reminderTime)) {
                reminderText.setText("No Reminder Added");
            } else {
                reminderText.setText("Reminder is set to " + reminderTime);
            }
        }
        btnRemindMe.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (TextUtils.isEmpty(noteText.getText().toString().trim())) {
                    noteText.setError("this field is required");
                } else {
                    createDatePicker();
                }
            }
        });
        noteText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                Log.d(TAG, "beforeTextChanged: ");
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                Log.d(TAG, "onTextChanged: ");
            }

            @Override
            public void afterTextChanged(Editable s) {
                if(TextUtils.isEmpty(s)){
            noteText.setHint("Enter your note here");
                }
            }
        });
    }

    /*this method create notification id to send the broadcast, this id is also added in the database and
     retrieved in the main activity while deleting the note */
    private String createNotificationId() {
        Random randomNumber = new Random();
        int uniqueRandomNumber = randomNumber.nextInt(900000);
        return String.valueOf(uniqueRandomNumber);
    }

    //To get the time in milliseconds and this value is stored in the database
    public String getTimeInMillisconds() {
        Date date = new Date();
        long value = date.getTime();
        String returnTime = String.valueOf(value);
        return returnTime;
    }

    public void createDatePicker() {
        DatePickerDialog datePickerDialog = new DatePickerDialog(this, new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                month = month + 1;
                Log.d(TAG, "onDateSet: year " + year + " month " + month + " day " + dayOfMonth);
                createTimePicker(year, month, dayOfMonth);
            }
        }, Calendar.getInstance().get(Calendar.YEAR), Calendar.getInstance().get(Calendar.MONTH), Calendar.getInstance().get(Calendar.DAY_OF_MONTH));
        datePickerDialog.getDatePicker().setMinDate(System.currentTimeMillis() - 1000);
        datePickerDialog.show();
    }

    public void createTimePicker(int year, int month, int dayOfMonth) {
        formattedDate = year + "/" + month + "/" + dayOfMonth;
        TimePickerDialog timePickerDialog = new TimePickerDialog(this, new TimePickerDialog.OnTimeSetListener() {
            @Override
            public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
                Log.d(TAG, "onTimeSet: hour " + hourOfDay + " minute " + minute);
                formattedDate = formattedDate + " " + hourOfDay + ":" + minute;
                Log.d(TAG, "onTimeSet: formatted date " + formattedDate);
                sendNotification(formattedDate);

            }
        }, Calendar.getInstance().get(Calendar.HOUR_OF_DAY), Calendar.getInstance().get(Calendar.MINUTE), true);
        timePickerDialog.show();
    }

    //call to the notification class that extends Broadcast receiver to display the notification
    public void sendNotification(String formattedDate) {
        //To make sure user has saved the note before adding reminder
        if (TextUtils.isEmpty(notificationID)) {
            Toast.makeText(this, "Kindly save the Note before adding the reminder", Toast.LENGTH_LONG).show();
        } else {
            if (convertTo24hrsMilliseconds(formattedDate ) > System.currentTimeMillis()) {
                Intent intent = new Intent(NoteActivity.this, Notification.class);
                if (TextUtils.isEmpty(notes)) {
                    notes = noteText.getText().toString().trim();
                }
                intent.putExtra("notes", notes);
                intent.putExtra("position", notificationID);
                PendingIntent pendingIntent = PendingIntent.getBroadcast(getApplicationContext(), Integer.parseInt(notificationID), intent, PendingIntent.FLAG_UPDATE_CURRENT);
                AlarmManager alarmManager = (AlarmManager) getSystemService(ALARM_SERVICE);
                alarmManager.set(AlarmManager.RTC_WAKEUP, convertTo24hrsMilliseconds(formattedDate), pendingIntent);
                reminderTime = convertToReadableDateFormat(formattedDate);
                reminderText.setText("Reminder is set to " + reminderTime);
                Toast.makeText(getApplicationContext(), "Reminder Added on " + convertToReadableDateFormat(formattedDate), Toast.LENGTH_LONG).show();
            } else {
                Toast.makeText(this, "The time is set to past time an so reminder cannot be set, Change it to future time to get Notification", Toast.LENGTH_LONG).show();
            }

        }
    }

    //for displaying the date and time in toast
    private String convertToReadableDateFormat(String formattedDate) {
        SimpleDateFormat dateFormat = new SimpleDateFormat("EEE, d MMM yyyy, h:mm a");
        Date date = new Date(formattedDate);
        formattedDate = dateFormat.format(date);
        return formattedDate;
    }

    public long convertTo24hrsMilliseconds(String formattedDate) {
        String myDate = formattedDate;
        long milliseconds = 1;
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy/MM/dd HH:mm");
        try {
            Date date = simpleDateFormat.parse(myDate);
            milliseconds = date.getTime();
            Log.d(TAG, "convertToMilliseconds: " + milliseconds);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return milliseconds;
    }

    //converts the Date and Time content got from Date and time picker alertDialog
    private long convertToMilliseconds(String formattedDate) {
        String myDate = formattedDate;
        long milliseconds = 1;
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy/mm/dd hh:mm");
        try {
            Date date = simpleDateFormat.parse(myDate);
            milliseconds = date.getTime();
            Log.d(TAG, "convertToMilliseconds: " + milliseconds);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return milliseconds;
    }

    @Override
    public void onBackPressed() {

        if(!noteText.getText().toString().equals(notes)){
            final AlertDialog.Builder backPressedAlertDialog = new AlertDialog.Builder(this);
            backPressedAlertDialog.setTitle("Note is not Saved");
            backPressedAlertDialog.setMessage("Do you want to continue without saving, The changes made will not be reflected");
            backPressedAlertDialog.setPositiveButton("No", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {

                }
            });
            backPressedAlertDialog.setNegativeButton("Yes", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    Intent backToMainActivity = new Intent(NoteActivity.this, MainActivity.class);
                    backToMainActivity.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(backToMainActivity);
                }
            });
            backPressedAlertDialog.show();

        }
        else{
            super.onBackPressed();
        }
    }
}

File: app/src/main/java/com/example/farith/dailynotes/ui/NoteActivity.kt
package com.example.farith.dailynotes.ui

import android.app.*
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.example.farith.dailynotes.R
import com.example.farith.dailynotes.databinding.ActivityNoteBinding
import com.example.farith.dailynotes.modelClass.NoteClass
import com.example.farith.dailynotes.receivers.Notification
import com.example.farith.dailynotes.ui.viewModels.MainViewModel
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class NoteActivity : AppCompatActivity() {
    private var formattedDate: String? = null
    private var notificationID: Int? = null
    private var reminderTime: String? = "No reminder Added"
    private var notes: String? = null
    private lateinit var mainViewModel: MainViewModel
    private lateinit var activityNoteBinding: ActivityNoteBinding
    private var createNew: Boolean = true;

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activityNoteBinding = DataBindingUtil.setContentView(this, R.layout.activity_note)

        bindData()
        setupViewModel()
    }

    private fun setupViewModel() {
        mainViewModel = ViewModelProvider(this)[MainViewModel::class.java]
    }

    private fun bindData() {
        intent?.let {
            if (!it.getBooleanExtra("ADD_BUTTON", true)) {
                createNew = false;
                val noteClass = it.getSerializableExtra("NOTE_CLASS") as NoteClass
                notes = noteClass.noteContent
                reminderTime = noteClass.reminderTime
                notificationID = noteClass.notificationId
                activityNoteBinding.noteText.setText(notes)
                //This line is to take the cursor to the end of the edit text field
                activityNoteBinding.noteText.setSelection(activityNoteBinding.noteText.text.toString().length)
                if (TextUtils.isEmpty(reminderTime)) {
                    activityNoteBinding.reminderText.text = "No Reminder Added"
                } else {
                    activityNoteBinding.reminderText.text = "Reminder is set to $reminderTime"
                }
            } else {
                createNew = true
            }

        }

        activityNoteBinding.saveBtn.setOnClickListener {
            notes = activityNoteBinding.noteText.text.toString().trim()
            if (TextUtils.isEmpty(notes)) {
                Toast.makeText(it.context, "Please enter some note.", Toast.LENGTH_SHORT).show()
                return@setOnClickListener;
            }
            val currentTime = getCurrentTime()
            val value = if (createNew) {
                val notificationId = currentTime.hashCode()
                NoteClass(
                    noteContent = notes, date = currentTime,
                    notificationId = notificationId, reminderTime = "0"
                )
            } else {
                NoteClass(
                    noteContent = notes, date = currentTime,
                    notificationId = notificationID, reminderTime = "0"
                )
            }
            mainViewModel.insertNotes(value)
            finish()

        }

        activityNoteBinding.btnRemindMe.setOnClickListener {
            if (TextUtils.isEmpty(
                    activityNoteBinding.noteText.text.toString().trim()
                )
            ) {
                activityNoteBinding.noteText.error = "this field is required"
            } else {
                createDatePicker()
            }
        }
        activityNoteBinding.noteText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(
                s: CharSequence,
                start: Int,
                count: Int,
                after: Int
            ) {
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            }

            override fun afterTextChanged(s: Editable) {
                if (TextUtils.isEmpty(s)) {
                    activityNoteBinding.noteText.hint = "Enter your note here"
                }
            }
        })

    }

    private fun getCurrentTime(): String {
        val date = Calendar.getInstance().time
        val dateFormat = SimpleDateFormat("EEE, d MMM yyyy, h:mm a")
        return dateFormat.format(date).toString()
    }

    private fun createDatePicker() {
        val datePickerDialog = DatePickerDialog(
            this,
            { _, year, month, dayOfMonth ->
                var month = month
                month += 1
                createTimePicker(year, month, dayOfMonth)
            },
            Calendar.getInstance()[Calendar.YEAR],
            Calendar.getInstance()[Calendar.MONTH],
            Calendar.getInstance()[Calendar.DAY_OF_MONTH]
        )
        datePickerDialog.datePicker.minDate = System.currentTimeMillis() - 1000
        datePickerDialog.show()
    }

    private fun createTimePicker(year: Int, month: Int, dayOfMonth: Int) {
        formattedDate = "$year/$month/$dayOfMonth"
        val timePickerDialog = TimePickerDialog(
            this,
            { view, hourOfDay, minute ->
                formattedDate = "$formattedDate $hourOfDay:$minute"
                sendNotification(formattedDate ?: "")
            },
            Calendar.getInstance()[Calendar.HOUR_OF_DAY],
            Calendar.getInstance()[Calendar.MINUTE],
            true
        )
        timePickerDialog.show()
    }

    //call to the notification class that extends Broadcast receiver to display the notification
    private fun sendNotification(formattedDate: String) {
        //To make sure user has saved the note before adding reminder
        if (notificationID == null) {
            Toast.makeText(
                this,
                "Kindly save the Note before adding the reminder",
                Toast.LENGTH_LONG
            ).show()
        } else {
            if (convertTo24hrsMilliseconds(formattedDate) > System.currentTimeMillis()) {
                val intent = Intent(this@NoteActivity, Notification::class.java)
                if (TextUtils.isEmpty(notes)) {
                    notes = activityNoteBinding.noteText.text.toString().trim { it <= ' ' }
                }
                intent.putExtra("notificationId", notificationID)
                intent.putExtra("notes", notes)
                val pendingIntent: PendingIntent =
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        PendingIntent.getBroadcast(
                            applicationContext,
                            notificationID!!.toInt(),
                            intent,
                            PendingIntent.FLAG_IMMUTABLE
                        )
                    } else {
                        PendingIntent.getBroadcast(
                            applicationContext,
                            notificationID!!.toInt(),
                            intent,
                            PendingIntent.FLAG_UPDATE_CURRENT
                        )
                    }

                val alarmManager = getSystemService(ALARM_SERVICE) as AlarmManager
                alarmManager[AlarmManager.RTC_WAKEUP, convertTo24hrsMilliseconds(formattedDate)] =
                    pendingIntent
                reminderTime = convertToReadableDateFormat(formattedDate)
                activityNoteBinding.reminderText.text = "Reminder is set to $reminderTime"
                Toast.makeText(
                    applicationContext,
                    "Reminder Added on " + convertToReadableDateFormat(formattedDate),
                    Toast.LENGTH_LONG
                ).show()
            } else {
                Toast.makeText(
                    this,
                    "The time is set to past time an so reminder cannot be set, Change it to future time to get Notification",
                    Toast.LENGTH_LONG
                ).show()
            }
        }
    }

    //for displaying the date and time in toast
    private fun convertToReadableDateFormat(formattedDate: String): String {
        var formattedDate = formattedDate
        val dateFormat = SimpleDateFormat("EEE, d MMM yyyy, h:mm a")
        val date = Date(formattedDate)
        formattedDate = dateFormat.format(date)
        return formattedDate
    }

    private fun convertTo24hrsMilliseconds(formattedDate: String): Long {
        val myDate = formattedDate
        var milliseconds: Long = 1
        val simpleDateFormat = SimpleDateFormat("yyyy/MM/dd HH:mm", Locale.getDefault())
        try {
            val date: Date? = simpleDateFormat.parse(myDate)
            milliseconds = date?.time ?: -1
        } catch (e: ParseException) {
            e.printStackTrace()
        }
        return milliseconds
    }

    //converts the Date and Time content got from Date and time picker alertDialog
    private fun convertToMilliseconds(formattedDate: String): Long {
        val myDate = formattedDate
        var milliseconds: Long = 1
        val simpleDateFormat = SimpleDateFormat("yyyy/mm/dd hh:mm", Locale.getDefault())
        try {
            val date = simpleDateFormat.parse(myDate)
            milliseconds = date.time
        } catch (e: ParseException) {
            e.printStackTrace()
        }
        return milliseconds
    }

    override fun onBackPressed() {
        if (activityNoteBinding.noteText.text.toString() != notes) {
            val backPressedAlertDialog = AlertDialog.Builder(this)
            backPressedAlertDialog.setTitle("Note is not Saved")
            backPressedAlertDialog.setMessage("Do you want to continue without saving, The changes made will not be reflected")
            backPressedAlertDialog.setPositiveButton("No") { dialog, which -> dialog.dismiss() }
            backPressedAlertDialog.setNegativeButton("Yes") { dialog, which ->
                val backToMainActivity =
                    Intent(this@NoteActivity, MainActivity::class.java)
                backToMainActivity.flags =
                    Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK
                startActivity(backToMainActivity)
            }
            backPressedAlertDialog.show()
        } else {
            super.onBackPressed()
        }
    }
}